### PR TITLE
fix(token): preserve absolute lifecycle.end across mint round-trip

### PR DESCRIPTION
**Pull request by starknetdev**
## Summary

The packed token stores `[minted_at, start_delay, end_delay]` where both delays are non-negative `u32`s, so `lifecycle.start >= minted_at` is a structural invariant — there is no representation for a window that opened before mint time.

The mint logic clamped `start_delay` to 0 when `start <= current_time`, but still computed `end_delay = lifecycle.end - lifecycle.start` from the **unclamped** start. The reconstructed `lifecycle.end` therefore landed at `mint_time + (end - start)` instead of the caller's intended `lifecycle.end`, silently extending the token's playable window past the cutoff the caller specified.

### Cases that hit this

| Caller passes | Old behavior | Fixed behavior |
|---|---|---|
| `start = abs_game_start, end = abs_game_end`, mint during live phase (`start < now`) | reconstructed end = `mint_time + (end - start)` (past cutoff) | reconstructed end = `end` ✓ |
| `start = None, end = abs_timestamp` (start defaults to 0) | `end_delay = end - 0` — overflows the 25-bit field, encodes wrong window | reconstructed end = `end` ✓ |
| `end != 0 && end <= current_time` (already expired) | silently `end_delay = 0` → token lives forever | panics with `MinigameToken: Lifecycle end must be in the future` |

### Fix

Clamp `effective_start = max(lifecycle.start, current_time)` and derive `end_delay` from the clamped value. The token's playable window for retroactive/late mints becomes `[mint_time, end]` — which matches reality, since the token didn't exist during the part of the window that already passed. Caller's `end` is preserved exactly, regardless of when the mint happens.

Also added a `lifecycle.end > current_time` precondition (when `end != 0`) so callers fail loudly on already-expired windows instead of getting a silently-immortal token.

### Caller impact

This is a behavior change for callers that were relying on the buggy round-trip:

- Previously, passing `start = past_T, end = past_T + N` and minting at `now` produced a token expiring at `now + N` (duration encoding). After the fix, this panics — the absolute `end` is in the past. Callers that wanted a token playable for `N` seconds from mint should pass `start = None, end = now + N` (or `start = now, end = now + N`).
- Multiple callers in the wider ecosystem (e.g. `budokan` tournament entries, `ticket_booth_component` — which got a similar caller-side fix in [a8efe5b](https://github.com/Provable-Games/game-components/commit/a8efe5b)) had to pre-clamp themselves to work around the broken round-trip. After this fix they no longer need to.

## Test plan

- [x] Existing test `test_core_token_playability_with_lifecycle` updated — it had been documenting the bug (passed `end = PAST_TIME + 1` and asserted the token expired at `mint_time + 1`). Now uses absolute `end = CURRENT_TIME + 1`.
- [x] New regression test `test_mint_with_start_in_past_clamps_to_mint_time_and_preserves_end`: mints with `start = PAST_TIME, end = FUTURE_TIME` mid-window, asserts reconstructed `lifecycle.start == mint_time` and `lifecycle.end == FUTURE_TIME`.
- [x] New regression test `test_mint_with_unset_start_uses_mint_time_as_start`: mints with `start = None, end = FUTURE_TIME`, asserts reconstructed `lifecycle.start == mint_time` and `lifecycle.end == FUTURE_TIME`.
- [ ] CI to run the full `embeddable_game_standard` test suite (local environment OOMs on test build).

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive tests for token lifecycle edge cases, including scenarios with past start times and unset start times during minting.

* **Bug Fixes**
  * Strengthened lifecycle validation to enforce that non-zero end times are in the future at mint time.
  * Refined delay calculations to properly handle start times in the past, ensuring correct playable window timing.

### CODE REVIEW

File: packages/embeddable_game_standard/src/token/token_component.cairo [hunk 1]
```diff
@@ -858,14 +858,41 @@ pub mod CoreTokenComponent {
...
-            let start_delay: u32 = if lifecycle.start > current_time {
-                (lifecycle.start - current_time).try_into().unwrap()
+            // A non-zero `end` must be in the future at mint time. Otherwise the
+            // packed `end_delay` would collapse to 0 — which the storage format
+            // treats as "no expiration" — silently producing a token that lives
+            // forever instead of one that's already expired (almost certainly a
+            // caller bug).
+            assert!(
+                lifecycle.end == 0 || lifecycle.end > current_time,
+                "MinigameToken: Lifecycle end must be in the future",
```
[the remark this answers is not in the document]
Comment (starknetdev) [reply packages/embeddable_game_standard/src/token/token_component.cairo]: Good catch — `start == end > current_time` slips through the original assertion (`validate()` only enforces `start <= end`) and falls into the `end_delay = 0` branch downstream, producing exactly the silent-immortal-token mode this PR is supposed to eliminate. Strengthened the condition to `lifecycle.end == 0 || (lifecycle.end > current_time && lifecycle.end > lifecycle.start)` and added a `#[should_panic]` test covering `start == end` (`test_mint_with_zero_length_window_panics`) in 2f8a178.

File: packages/embeddable_game_standard/src/token/token_component.cairo [hunk 2]
```diff
@@ -858,14 +858,41 @@ pub mod CoreTokenComponent {
...
+            );
```
[the remark this answers is not in the document]
Comment (starknetdev) [reply packages/embeddable_game_standard/src/token/token_component.cairo:878]: Skipping this one — every other `assert!` in `token_component.cairo` uses inline strings (e.g. `"MinigameToken: token_ids array cannot be empty"` repeated 12+ times, plus `"MinigameToken: Player name is empty"` etc.). I checked the project's `CLAUDE.md` / `AGENTS.md` and didn't find the cited "all error messages must be constants" guideline, and there's no shared error-constants module under `packages/embeddable_game_standard/src/`. Switching just this one to a constant would be inconsistent with the surrounding code. Happy to revisit if you'd prefer a centralized refactor — that'd be its own PR sweeping all the inline messages.

**Comment by starknetdev** [comment]
Pushed [2f8a178](https://github.com/Provable-Games/game-components/pull/110/commits/2f8a178) addressing review feedback:

- **codecov / Codex / CodeRabbit (missing coverage on the panic branch):** added three `#[should_panic]` regression tests in `test_core_token.cairo` covering `end < current_time`, `end == current_time`, and `start == end > current_time`. Should bring patch coverage to 100%.
- **Gemini (assertion missed `start == end`):** strengthened the precondition to `lifecycle.end == 0 || (lifecycle.end > current_time && lifecycle.end > lifecycle.start)`. The earlier version slipped a zero-length window past the guard, which then fell into `end_delay = 0` downstream — same silent-immortal-token failure mode this PR was supposed to eliminate. Replied on the inline thread.
- **CodeRabbit (use named error constant):** declined — every other `assert!` in this file uses inline strings (12+ occurrences of `"MinigameToken: token_ids array cannot be empty"` alone), and the cited "all error messages must be constants" guideline isn't actually in the project's `CLAUDE.md` / `AGENTS.md`. Replied on the inline thread.
